protoc-gen-twirp/generator.go

```diff
@@ -591,15 +591,26 @@ func (t *twirp) generateUtils() {
 	t.P(`// error response (status 500), and error hooks are properly called with the panic wrapped as an error.`)
 	t.P(`// The panic is re-raised so it can be handled normally with middleware.`)
 	t.P(`func ensurePanicResponses(ctx `, t.pkgs["context"], `.Context, resp `, t.pkgs["http"], `.ResponseWriter, hooks *`, t.pkgs["twirp"], `.ServerHooks) {`)
-	t.P(`  if r := recover(); r != nil {`)
-	t.P(`    // Wrap the panic as an error so it can be passed to error hooks.`)
-	t.P(`    // The original error is accessible from error hooks, but not visible in the response.`)
-	t.P(`    err := errFromPanic(r)`)
-	t.P(`    twerr := &internalWithCause{msg: "Internal service panic", cause: err}`)
-	t.P(`    writeError(ctx, resp, twerr, hooks)`)
+	t.P(`	if r := recover(); r != nil {`)
+	t.P(`		// Wrap the panic as an error so it can be passed to error hooks.`)
+	t.P(`		// The original error is accessible from error hooks, but not visible in the response.`)
+	t.P(`		err := errFromPanic(r)`)
+	t.P(`		twerr := &internalWithCause{msg: "Internal service panic", cause: err}`)
+	t.P(`		// Set Content-Length of response to avoid switching to chunked transfer`)
+	t.P(`		// encoding when we flush, and so the client knows that it has received the`)
+	t.P(`		// complete response when we flush.`)
+	t.P(`		length := len(marshalErrorToJSON(twerr))`)
```

Comment:
We end up calling `marshalErrorToJSON` twice in this code path - once here, and once in `writeError`. I could refactor to avoid that, but wanted the first revision to be clear, so I left in the double marshal.

Reply:
I think marshaling twice should be fine. Most error responses are small, most panics are unfrequent. If we ever find performance issues with this we can always refactor later.

Reply:
Why don't we send Content-Length on every error response? This could all move inside `writeError`, which currently calls `resp.WriteHeader` right before calling `marshalErrorToJSON`.

Why don't we send Content-Length on successful responses too? In those cases we definitely have the bytes ready before we write the header (so we can report any marshal errors).

Reply:
That's a good point, @rhysh. I like that. I'll revise to always set Content-Lengths.